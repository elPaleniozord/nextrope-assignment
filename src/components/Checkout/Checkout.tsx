import { SyntheticEvent, useState } from "react"
import { Button, Container, Form, Spinner } from "react-bootstrap"
import { Navigate } from "react-router-dom"
import { useAppSelector } from "../../app/hooks"
import { selectCart } from "../../features/cart/cartSlice"
import AlertPopup from "../Alert/Alert"

interface ReqStatus {
  pending: boolean
  message: string | undefined
}

const Checkout = () => {
  const [state, setState] = useState({
    name: '',
    surname: '',
    town: '',
    postalCode: ''
  })
  const [valid, setValid] = useState(false)
  const [status, setStatus] = useState<ReqStatus>({
    pending: false, 
    message: undefined
  })
  const cart = useAppSelector(selectCart)

  const handleFormSubmit = (e: SyntheticEvent) => {
    e.preventDefault()
    e.stopPropagation()

    if(!valid) return
    else {
      const order = {
        order: cart.items.map(item => ({id: item.id, quantity: item.amount})),
        first_name: state.name, last_name: state.name, city: state.town, zip_code: state.postalCode
      }
      //disable form controls while request is pending
      setStatus({...status, pending: true})

      fetch('http://localhost:3001/api/order', {
        method: 'POST',
        body: JSON.stringify(order),
        headers: { 'Content-Type': 'application/json' }
      }).then(res => {
        if(res.ok) { 
          setStatus({pending: false, message: 'success'})
        } else {
          //task failed successfully
          setStatus({pending: false, message: 'Błędny format zamówienia'})
        }
      }).catch(err => {
        setStatus({pending: false, message: 'Błąd serwera '})
      })
    }
  }

  const handleInputChange = (e: React.SyntheticEvent) => {
    const target = e.target as HTMLInputElement
    const regex = new RegExp(target.pattern)
    setState({...state, [target.id]: target.value})

    if(target.value.match(regex)) setValid(true)
    else setValid(false)
  }

  if(cart.total === 0 || cart.items.length === 0) return <Navigate to='/' replace={true}  />

  return (
    <Container>
      <h2>Podumowanie</h2>
      <Form validated={valid} onSubmit={handleFormSubmit}>
        <h3>Dane do wysyłki</h3>
        <fieldset disabled={status.pending || status.message === 'success'}>
          <Form.Group controlId="name">
            <Form.Label>Imię</Form.Label>
            <Form.Control required type='text' minLength={3} pattern="^[A-Za-z]+$" placeholder="Imię" onChange={handleInputChange} value={state.name}/>
            <Form.Control.Feedback type='invalid'>Imię powinno być dłuższe niż 2 litery oraz zawierać tylko litery alfabetu</Form.Control.Feedback>
          </Form.Group>

          <Form.Group controlId="surname">
            <Form.Label>Nazwisko</Form.Label>
            <Form.Control required type='text' minLength={2} pattern="^[A-Za-z]+$" placeholder="Nazwisko" onChange={handleInputChange} value={state.surname}></Form.Control>
            <Form.Control.Feedback type='invalid'>Nazwisko powinno być dłuższe niż 2 litery oraz zawierać tylko litery alfabetu</Form.Control.Feedback>
          </Form.Group>

          <Form.Group controlId="town">
            <Form.Label>Miejscowość</Form.Label>
            <Form.Control required type='text' minLength={2} placeholder="Miejscowość" pattern="^[a-zA-Z0-9-\s]+$" onChange={handleInputChange} value={state.town}></Form.Control>
            <Form.Control.Feedback type='invalid'>Nazwa miejscowości powinna składać się z co najmniej 2 znaków alfanumerycznych</Form.Control.Feedback>
          </Form.Group>

          <Form.Group controlId="postalCode">
            <Form.Label>Kod pocztowy</Form.Label>
            <Form.Control required type='postalCode' pattern="^[0-9]{2}-[0-9]{3}$" placeholder="dd-ddd" onChange={handleInputChange} value={state.postalCode}></Form.Control>
            <Form.Text></Form.Text>
            <Form.Control.Feedback type="invalid">Poprawny format kodu pocztowego to dd-ddd (d oznacza cyfrę)</Form.Control.Feedback>
          </Form.Group>
        </fieldset>

        <Button style={{width: '100%', marginTop: "2rem"}} type='submit' disabled={status.pending || status.message === 'success'}>
          {status.pending ? (<Spinner animation="border" />) : 'ZAMAWIAM I PŁACĘ'}
        </Button>
      </Form>

      <AlertPopup status={status.message}></AlertPopup>
    </Container>
  )
}

export default Checkout