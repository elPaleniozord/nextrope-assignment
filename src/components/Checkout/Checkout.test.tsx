import { configureStore } from '@reduxjs/toolkit';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import cartReducer, { addProduct } from '../../features/cart/cartSlice';
import Checkout from './Checkout';

const book = {
  "id": 457,
  "title": "Matematyka 1. Podręcznik. Zakres podstawowy",
  "author": "M. Karpiński, M. Dobrowolska, M. Braun, J. Lech",
  "cover_url": "http://localhost:3001/static/cover/book/457.jpg",
  "pages": 280,
  "price": 3200,
  "currency": "PLN"
}

export function renderWithStore (
  ui, {
    preloadedState, 
    store = configureStore({reducer: {cart: cartReducer}, preloadedState}),
    ...renderOptions
  } = {}
) {
  function Wrapper({children}) {
    store.dispatch(addProduct(book))
    return <Provider store={store}>{children}</Provider>
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions})
}

test('should render checkout component', () => {
  const {asFragment} = renderWithStore(
    <MemoryRouter>
      <Checkout />
    </MemoryRouter>
  )
  expect(asFragment()).toMatchSnapshot()
})