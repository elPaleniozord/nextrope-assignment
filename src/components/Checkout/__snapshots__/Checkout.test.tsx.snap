// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render checkout component 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h2>
      Podumowanie
    </h2>
    <form
      class=""
    >
      <h3>
        Dane do wysyłki
      </h3>
      <fieldset>
        <div>
          <label
            class="form-label"
            for="name"
          >
            Imię
          </label>
          <input
            class="form-control"
            id="name"
            minlength="3"
            pattern="^[A-Za-z]+$"
            placeholder="Imię"
            required=""
            type="text"
            value=""
          />
          <div
            class="invalid-feedback"
          >
            Imię powinno być dłuższe niż 2 litery oraz zawierać tylko litery alfabetu
          </div>
        </div>
        <div>
          <label
            class="form-label"
            for="surname"
          >
            Nazwisko
          </label>
          <input
            class="form-control"
            id="surname"
            minlength="2"
            pattern="^[A-Za-z]+$"
            placeholder="Nazwisko"
            required=""
            type="text"
            value=""
          />
          <div
            class="invalid-feedback"
          >
            Nazwisko powinno być dłuższe niż 2 litery oraz zawierać tylko litery alfabetu
          </div>
        </div>
        <div>
          <label
            class="form-label"
            for="town"
          >
            Miejscowość
          </label>
          <input
            class="form-control"
            id="town"
            minlength="2"
            pattern="^[a-zA-Z0-9-\\\\s]+$"
            placeholder="Miejscowość"
            required=""
            type="text"
            value=""
          />
          <div
            class="invalid-feedback"
          >
            Nazwa miejscowości powinna składać się z co najmniej 2 znaków alfanumerycznych
          </div>
        </div>
        <div>
          <label
            class="form-label"
            for="postalCode"
          >
            Kod pocztowy
          </label>
          <input
            class="form-control"
            id="postalCode"
            pattern="^[0-9]{2}-[0-9]{3}$"
            placeholder="dd-ddd"
            required=""
            type="postalCode"
            value=""
          />
          <small
            class="form-text"
          />
          <div
            class="invalid-feedback"
          >
            Poprawny format kodu pocztowego to dd-ddd (d oznacza cyfrę)
          </div>
        </div>
      </fieldset>
      <button
        class="btn btn-primary"
        style="width: 100%; margin-top: 2rem;"
        type="submit"
      >
        ZAMAWIAM I PŁACĘ
      </button>
    </form>
  </div>
</DocumentFragment>
`;
